free/src/main/scala/cats/free/FreeT.scala

```diff
@@ -0,0 +1,263 @@
+    liftT(a).flatMap({
+      case Xor.Left(a) => pure(a)
+      case Xor.Right(s) => roll(s)
+    })
+
+  def tailRecM[S[_], M[_]: Applicative, A, B](a: A)(f: A => FreeT[S, M, A Xor B]): FreeT[S, M, B] =
+    f(a).flatMap {
+      case Xor.Left(a0) => tailRecM(a0)(f)
+      case Xor.Right(b) => pure[S, M, B](b)
+    }
+
+  def liftT[S[_], M[_], A](value: M[A])(implicit M: Functor[M]): FreeT[S, M, A] =
+    Suspend(M.map(value)(Xor.left))
+
+  /** A version of `liftT` that infers the nested type constructor. */
+  def liftTU[S[_], MA](value: MA)(implicit M: Unapply[Functor, MA]): FreeT[S, M.M, M.A] =
+    liftT[S, M.M, M.A](M.subst(value))(M.TC)
+
+  /** Suspends a value within a functor in a single step. Monadic unit for a higher-order monad. */
+  def liftF[S[_], M[_], A](value: S[A])(implicit M: Applicative[M]): FreeT[S, M, A] =
+    Suspend(M.pure(Xor.right(value)))
+
+  def roll[S[_], M[_], A](value: S[FreeT[S, M, A]])(implicit M: Applicative[M]): FreeT[S, M, A] =
+    liftF[S, M, FreeT[S, M, A]](value).flatMap(identity)
+
+}
+
+private[free] sealed trait FreeTInstances3 {
+  implicit def catsFreeMonadStateForFreeT[S[_], M[_], E](implicit M1: MonadState[M, E]): MonadState[FreeT[S, M, ?], E] =
+    new MonadState[FreeT[S, M, ?], E] with FreeTMonad[S, M] {
+      override def M = implicitly
+      override def get =
+        FreeT.liftT(M1.get)
+      override def set(s: E) =
+        FreeT.liftT(M1.set(s))
+    }
+}
+
+private[free] sealed trait FreeTInstances2 extends FreeTInstances3 {
+  implicit def catsFreeMonadErrorForFreeT[S[_], M[_]: RecursiveTailRecM, E](implicit E: MonadError[M, E]): MonadError[FreeT[S, M, ?], E] =
+    new MonadError[FreeT[S, M, ?], E] with FreeTMonad[S, M] {
+      override def M = implicitly
+      override def handleErrorWith[A](fa: FreeT[S, M, A])(f: E => FreeT[S, M, A]) =
+        FreeT.liftT[S, M, FreeT[S, M, A]](E.handleErrorWith(fa.toM)(f.andThen(_.toM)))(M).flatMap(identity)
+      override def raiseError[A](e: E) =
+        FreeT.liftT(E.raiseError[A](e))(M)
+    }
+}
+
+private[free] sealed trait FreeTInstances1 extends FreeTInstances2 {
+  implicit def catsFreeFlatMapForFreeT[S[_], M[_]](implicit M0: Applicative[M]): FlatMap[FreeT[S, M, ?]] =
+    new FreeTFlatMap[S, M] {
+      implicit def M: Applicative[M] = M0
+    }
+
+  implicit def catsFreeTransLiftForFreeT[S[_]]: TransLift.Aux[FreeT[S, ?[_], ?], Functor] =
+    new TransLift[FreeT[S, ?[_], ?]] {
+
+      type TC[M[_]] = Functor[M]
+
+      override def liftT[M[_]: Functor, A](ma: M[A]): FreeT[S, M, A] =
+        FreeT.liftT(ma)
+    }
+}
+
+private[free] sealed trait FreeTInstances0 extends FreeTInstances1 {
+  implicit def catsFreeMonadForFreeT[S[_], M[_]](implicit M0: Applicative[M]): Monad[FreeT[S, M, ?]] with RecursiveTailRecM[FreeT[S, M, ?]] =
+    new FreeTMonad[S, M] {
+      def M = M0
+    }
+
+  implicit def catsFreeCombineForFreeT[S[_], M[_]: Applicative: RecursiveTailRecM: SemigroupK]: SemigroupK[FreeT[S, M, ?]] =
```

Comment:
I don't think you need `RecursiveTailRecM` unless you also have `FlatMap` or `Monad` and I don't see that here.